[Week3] 3차 필수, 심화과제

nahy-512 opened the pull request:
## 📮 관련 이슈
- closed #6 

## 📌 작업 내용
<3차 과제>
- Home, SignUp 부분을 MVVM 패턴으로 바꿨어요. (+ SignIn)
- 필수, 심화과제 애니메이션 구현했어요.

<기타 리팩토링>
- 마이페이지 로그아웃 기능 추가했어요.
- 마이페이지 UI 좀 수정하고, 컴포넌트화 시켜봤어요.
- 로그인 유효성 확인 로직을 캡슐화 해봤어요.
- 자동로그인을 위한 UserPreferences를 class -> object로 변경, Application 클래스 생성 후 ApplicationContext를 넣어주는 방식으로 변경

## 📸 스크린샷
| 스크린샷 |
|:--:|
| <video width="360" src="https://github.com/user-attachments/assets/af244e3d-5a22-4d71-b6ed-2f86621d6ff6" /> |

## 😅 미구현
- [ ] Coil 적용해보고 싶었으나 실패
- [x] SignIn도 MVVM 구조로 바꾸기
- [ ] Splash 자동로그인 이슈 확인

## 🫛 To. 리뷰어
- 커밋 컨벤션을 바꿔보앗서요. 드디어 심화과제 했다. 코멘트 많이많이 달아주세요~~
- 애니메이션 너무 어렵다.. 제가 과제를 잘 이해한 건지 모르겠지만 심화과제 이렇게 하는 거 맞는지 모르겠네요.
- damping ratio 구하는 공식이 따로 있다고 해서.. 함수로 추출을 해보았습니다.
- Search, MyPage는 아직 뭔가 뷰모델 쓸 필요 없을 것 같아서.. SignIn 부분만 추가로 MVVM으로 바꿔볼까 합니다. 여러분들은 어떻게 생각하시나요? -> SignIn MVVM 구조 적용 완
- ~~`HomeViewModel`에서 유저 정보를 불러올 때 context를 직접 주입받고 있는데.. 4주차에 hilt 써서 의존성 주입받는 형태로 바꿔볼게요.~~  -> ApplicationContext 활용 방식으로 우선 변경
---
- `SignUp` -> `SignIn` 이동 시 뷰모델에서 데이터 받게 리팩토링해뒀어요! 코드 확인해보시면 좋을 것 같습니다ㅎㅎ
- Home, SignUp, SignIn을 MVVM 구로조 변경한 커밋은 따로 나눠놨어요. 보시고 Route와 Screen 분리의 이유에 대해 감 잡아보시면 좋지 않을까 싶습니다.
- 화면이동, 토스트 표시 `SideEffect` 처리에 대해서는 `SignInRoute`에서 처리한 것처럼 라우트에서 그냥 처리를 할지, 아니면 다른 구조를 사용해볼지 고민이 좀 되네요ㅎㅎ 관리 방식에 대해서는 저도 조금 더 생각해보고 리팩토링 진행하겠습니다. 좋은 의견이 있다면 남겨주세요!!

nahy-512 committed:
chore: 뷰모델 의존성 추가

Related to: #6

nahy-512 committed:
refactor: 홈 화면 MVVM 구조로 변경

Related to: #6

nahy-512 committed:
design: 바텀바 디자인 수정

Related to: #6

nahy-512 committed:
feat: FlipCard 기능 구현

Related to: #6

nahy-512 committed:
style: 컨벤션 수정

Related to: #6

nahy-512 committed:
Merge remote-tracking branch 'origin/main' into week3

nahy-512 committed:
feat: 심화과제 카드 fliping 구현

Related to: #6

nahy-512 committed:
feat: dampingRatio 계산 로직 추가

Related to: #6

nahy-512 committed:
style: 기타 수정

- FlipCard 클릭 이벤트 수정
- MainBottombar 구분선 삭제

Related to: #6

nahy-512 committed:
refactor: 회원가입 MVVM 패턴 적용

Related to: #6

nahy-512 committed:
style: 마이페이지 컴포넌트 분리 및 디자인 수정

Related to: #6

nahy-512 committed:
style: 주석 추가

Related to: #6

nahy-512 committed:
refactor: collectAsState -> collectAsStateWithLifecycle

Related to: #6

nahy-512 committed:
style: 기타 수정

Related to: #6

nahy-512 committed:
refactor: UserPreferences 구조 변경

- Application 클래스 생성
- conntext 직접 주입 -> Application context 전달 방식으로 변경

Related to: #6, #2

nahy-512 committed:
refactor: 로그인 MVVM 패턴 적용, 에러 처리 로직 적용

- ViewModel에서 SavedStateHandle 처리
- 에러 처리 로직 캡슐화
- 유저 정보 저장 책임 분리

Related to: #6

nahy-512 committed:
feat: 마이페이지 로그아웃 로직 추가

- 로그아웃 시 prefs SharedPreferences 정보 삭제
- 로그아웃 후 로그인 화면으로 이동

Related to: #6

ckals413 commented on app/src/main/java/com/sopt/dive/data/local/UserPreferences.kt:
```diff
@@ -5,11 +5,23 @@ import android.content.SharedPreferences
 import com.sopt.dive.domain.model.auth.UserInfo
 
-class UserPreferences(context: Context) {
-    private val spf: SharedPreferences = context.getSharedPreferences(SPF_KEY, Context.MODE_PRIVATE)
+//TODO: hilt 사용
+object UserPreferences {
+    private const val SPF_KEY = "spf_key"
+    private const val USER_ID_KEY = "user_id"
+    private const val USER_PASSWORD_KEY = "user_password"
+    private const val USER_NICKNAME_KEY = "user_nickname"
+    private const val USER_MBTI_KEY = "user_mbti"
+    private const val IS_LOGGED_IN_KEY = "auto_login"
+
+    private lateinit var prefs: SharedPreferences
```
class를 object로 ,
lateinit으로 변수 선언을 바꾼이유가 뭔가요?

nahy-512 replied:
SharedPreference 접근을 위해서는 context가 필요한데요..!

이번에 MVVM 패턴으로 바꾸면서 뷰모델에서 context를 받아오려다가.. 이 방법은 너무 권장되지 않는 방법이겠다 싶어서 ApplicationContext를 받아오게끔 하려고 변경한 부분입니다!

각각의 변경 이유에 대해 더 자세히 설명하자면 아래와 같습니다.

### 1. class → object
SharedPreferences는 앱 전체에서 공용으로 접근해야 하는 저장소이기 때문에, 매번 인스턴스를 생성할 필요 없이 하나의 객체로 관리하는 것이 더 적절합니다.

저는 Application 클래스(DiveApplication)에서 `UserPreferences.init(context)`를 호출해뒀는데요!
이렇게 한 번 호출해두면 어디서든 동일한 인스턴스를 참조할 수 있어, 의존성 주입 없이도 간편하게 접근할 수 있습니다.

즉, object 전환으로 전역 상태 관리와 코드 일관성을 확보했다고 볼 수 있을 것 같습니다.

### 2.  prefs 지연 초기화 (lateinit)
Context 의존성을 제거하고, 초기화 시점을 명확히 분리하기 위해 lateinit을 사용했습니다.

SharedPreferences 생성에는 Context가 필요하지만,
ViewModel 등에서는 Context를 직접 참조하는 것이 메모리 누수나 테스트 어려움으로 이어질 수 있습니다.

그래서 init(context)를 통해 ApplicationContext를 한 번만 전달받아 초기화하고,
이후에는 context 없이 안전하게 접근할 수 있도록 했습니다.

ckals413 commented on app/src/main/java/com/sopt/dive/presentation/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+    val myProfile: StateFlow<UserInfo> = _myProfileInfo.asStateFlow()
+
+    private val _friendProfiles = MutableStateFlow(dummyFriendProfile)
+    val friendProfile: StateFlow<ImmutableList<FriendProfile>> = _friendProfiles
```
ImmutableList를 사용하신 이유가 궁금합니다! 사용시 장점이 있나요?

nahy-512 replied:
2차 과제에도 달아주셨던 코멘트인 걸로 기억해요! 제가 그때 답글을 늦게 남겨서ㅠㅠ

우선, 제가 여기에 `ImmutableList`를 사용한 이유는 UI 상태의 불변성을 보장하기 위해서입니다.

일반 `List`는 참조가 외부에 노출될 경우, 외부에서 원본 데이터를 변경할 위험이 있습니다.
반면 `ImmutableList`는 읽기 전용으로 보장되어 Compose에서 State가 변경되지 않았는데도 재조합이 일어나는 문제나 예기치 않은 데이터 변조를 방지할 수 있습니다.

즉, ViewModel에서 방출하는 데이터가 항상 안전하고 예측 가능한 상태(immutable) 로 유지되도록 하기 위해 사용했습니다ㅎㅎ

특히 Compose에서는 ImmutableList를 사용하면 리컴포지션 최적화에도 유리합니다.
List가 불변이면, Diff 검사가 명확해지고 불필요한 UI 갱신을 줄일 수도 있어요!

추가로, `Stable`과 `Immutable`에 대해서 공부해보시면 좋을 것 같아요.
skydoves님의 [Jetpack Compose 성능 최적화를 위한 Stability 이해하기](https://velog.io/@skydoves/compose-stability) 추천드립니다!

ckals413 commented on app/src/main/java/com/sopt/dive/presentation/home/HomeScreen.kt:
```diff
@@ -33,11 +34,11 @@ fun HomeRoute(
 
     viewModel.setMyProfile(context)
-    val userInfo = viewModel.myProfile.collectAsState().value
+    val userInfo = viewModel.myProfile.collectAsStateWithLifecycle().value
```
이 두개는 어떤 차이점이 있나요?

nahy-512 replied:
`collectAsState()`와 `collectAsStateWithLifecycle()`는 둘 다 Flow를 State로 변환한다는 점에서는 동일한데요,
둘의 차이는 Lifecycle 인식 여부에 있습니다.

`collectAsState()`는 단순히 Flow를 State로 변환해서 UI에 반영하지만,
컴포저블이 비활성화(ex. 화면이 백그라운드로 가거나, Lifecycle이 STOPPED 상태) 되어도 계속해서 Flow를 수집합니다.

반면 `collectAsStateWithLifecycle()`은 Lifecycle을 인식하여,
컴포넌트가 STARTED 상태일 때만 Flow를 수집하고,
비활성화되면 자동으로 수집을 중단합니다.

즉, `collectAsStateWithLifecycle()`을 사용하면
리소스 낭비나 불필요한 데이터 흐름을 방지하고, Lifecycle 안정성을 보장할 수 있습니다.

Jetpack Compose 공식 문서에서도 `Flow`를 UI에서 수집할 때는 `collectAsStateWithLifecycle()` 사용을 권장하고 있습니다!

ckals413 commented on app/src/main/java/com/sopt/dive/presentation/mypage/component/ProfileSpringCard.kt:
```diff
@@ -0,0 +1,179 @@
+    damping = 20f,
+    stiffness = stiffness,
+    mass = 1f
+)
```
이렇게 뺀거 좋은 것 같슴니다!

angryPodo replied:
상수로 추출한건 저도 좋다고 생각해요! 그런데 public으로 노출해야할 상수는 아닌것 같아요😅 접근제어자를 붙여주면 더 좋을 것 같네요!

dmp100 commented on app/src/main/java/com/sopt/dive/data/local/UserPreferences.kt:
```diff
@@ -5,11 +5,23 @@ import android.content.SharedPreferences
 
-class UserPreferences(context: Context) {
-    private val spf: SharedPreferences = context.getSharedPreferences(SPF_KEY, Context.MODE_PRIVATE)
+//TODO: hilt 사용
```
아직도 Hilt를 야매로 아는 저한테,, 친절한 설명 부탁드릴 수 있을까요 ???
(자료도 공유해주시면 감사하겠습니다 ~! )

nahy-512 replied:
음,, 일단 전 이번 과제에 힐트를 적용한 게 아니라 코드에 대해서는 추가로 설명드리기는 어려울 것 같아요.
이 주석은 나중에 의존성 주입 시에 hilt를 사용해서 해보려고 추후 구현 예정임을 TODO로 나타내본 거였어요ㅎㅎ

아마 4주차는 뷰모델 팩토리 써서 해보고, 이후 주차에 힐트를 써서 리팩토링해볼까 싶긴 합니다! 사용하게 된다면 공부한 자료 공유보겠습니다!!

dmp100 commented on app/src/main/java/com/sopt/dive/presentation/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,93 @@
+        _myProfile.value = UserPreferences.getUserInfo()
+    }
+
+    companion object {
```
차민님한테 코리단것과 비슷하게, `DummyData.kt`와 같이 따로 더미데이터로 관리하는거는 
어떻게 생각하시는지 궁금합니다 !
만약 실제 프로젝트라면 더미데이터를 따로 빼서 구분하는게 추후 API로 데이터받아오는게 유리한것인지 ?
아니면 더미데이터이기때문에 막(?) 써도 괜찮은것인지 ?

angryPodo replied:
저도 이 부분이 궁금하네요!
저라면 뷰모델 클래스의 동반 객체로 취급하지는 않을 것 같아요.
이렇게 되면 `HomeScreen`이 `HomeViewModel`의 내부구현에 강하게 의존한다고 생각해요.

별도의 파일로 분리하거나 현재 소스코드의 하단에 적어도 좋을 것 같아요ㅎㅎ
ViewModel은 `StateFlow`를 통해 데이터를 제공하는 역할에 충실하도록 설계해봅시다.

nahy-512 replied:
우왕 저는 해당 더미데이터를 서버에서 GET으로 데이터를 받아오고 나면 삭제할 임시 코드라고 생각하고 이외 부분까지는 별로 고려하지 않았었던 것 같습니다.

더미데이터를 임시로 볼 게 아니라, UI 테스트를 위해서라도 어디선가 저장할 필요성은 있을 것 같네요.
그렇게 되면 두분이 말씀하셨던 것처럼 별도의 파일로 분리하거나 현재 소스코드의 하단에 적는 식으로 관리하는 게 좋을 것 같아요 !

ckals413 reviewed (approved):
고생하셨습니다!! 
MVVM 패턴이 낮설어서 공부가 더 필요한데, 커밋을 따로 나눠주셔서 보기 너무 좋았습니다!! 

커밋 컨벤션을 바꾸셨는데 커밋 컨벤션의 기준이 있으신가요? (이슈 번호를 넣는다는지) 

"Search, MyPage는 아직 뭔가 뷰모델 쓸 필요 없을 것 같아서.."이 생각에 완전 동의합니다! 아직 로컬에서 데이터를 가져와서 지금 view모델을 만드는 게 오히려 보일러플레이트 코드가 늘어난다고 생각해요! 필요할 때 추가하는 방식이 좋아보입니다!

dmp100 reviewed (approved):
항상 흠잡을데가 없어서보여서 코리달때 어렵네요ㅜㅜ
이번에 Hilt를 적용하신것 같은데, 꽤 고수준의 기술을 보여주신것 같아서 항상 
잘 보고 배워요 !!

angryPodo commented on app/src/main/java/com/sopt/dive/data/local/UserPreferences.kt:
```diff
@@ -5,11 +5,23 @@ import android.content.SharedPreferences
-class UserPreferences(context: Context) {
-    private val spf: SharedPreferences = context.getSharedPreferences(SPF_KEY, Context.MODE_PRIVATE)
+//TODO: hilt 사용
+object UserPreferences {
+    private const val SPF_KEY = "spf_key"
+    private const val USER_ID_KEY = "user_id"
+    private const val USER_PASSWORD_KEY = "user_password"
+    private const val USER_NICKNAME_KEY = "user_nickname"
+    private const val USER_MBTI_KEY = "user_mbti"
+    private const val IS_LOGGED_IN_KEY = "auto_login"
+
+    private lateinit var prefs: SharedPreferences
+
+    fun init(context: Context) {
+        prefs = context.applicationContext.getSharedPreferences(SPF_KEY, Context.MODE_PRIVATE)
+    }
```
`class`에서 `object`로 변경하신 이유가 Hilt를 사용하지 않고 의존성을 주입하기 위함으로 보이네요!

좋은 시도입니다 😊

아마도 `UserPreferences`의 인스턴스를 모든 뷰모델이나 컴포저블에 `context`와 함께 주입하는 게 매우 번거롭기 때문에 Application 생명주기에 맞춰 초기화되는 전역 싱글톤을 선택하신 것 같아요.

이 선택이 왜 위험한지, 어떻게 개선할 수 있는지 제 관점에서 리뷰해 볼게요.

```kotlin
// DiveApplication.kt
class DiveApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        UserPreferences.init(this) // (1)
    }
}

// UserPreferences.kt
object UserPreferences {
    // ...
    private lateinit var prefs: SharedPreferences // (2)
    fun init(context: Context) { ... } // (3)
    // ...
}

```

`lateinit var`와 `init()` 패턴은 '시간적 결합’이라는 문제를 가지고 있어요.

1. `UserPreferences`의 모든 공개 함수는 `init()`이 **반드시 먼저** 호출되어야만 `prefs`에 접근할 수 있다.
2. 컴파일러는 이 순서를 강제하지 못한다.
3. 만약 `DiveApplication.onCreate()`가 실행되기 전에 (예: 테스트 코드, Content Provider) `UserPreferences.getUserInfo()`가 호출되면 앱은 `LateinitPropertyAccessException`와 함께 **런타임에 즉시 크래시한다.**

지금 코드는 "실행 순서를 잘 지킬 거다"라는 암묵적인 기대위에 설계 되었어요.

그리고 아래는 테스트 관점인데요

`HomeViewModel` 같은 클래스에서 `object`의 메서드를 직접(정적으로) 호출하고 있어요

```kotlin
// HomeViewModel.kt
fun loadMyProfileInfo() {
    _myProfile.value = UserPreferences.getUserInfo() // <-- 하드코딩
}
```

이 구조의 문제는 '단위 테스트'가 불가능해요. `loadMyProfileInfo()`를 테스트하려면 `UserPreferences.getUserInfo()`가 가짜 `UserInfo`를 반환하도록 만들어야 하는데, `object`는 Mocking이 매우 어렵거나 불가능해요(목적을 생각해보면 맞지 않아요)

결국 ViewModel 로직만 테스트하고 싶은데도, `SharedPreferences`의 실제 파일 I/O를 수행하는 비싼 테스트를 돌려야 해요.

---

자 실컷 문제만 지적하고 답이 없으면 답답하지요?ㅋ

그럼 생각해봅시다. 저는 이렇게 먼저 생각이 들어요.

> Hilt를 사용하지 않는 게 목적이면, Hilt가 해주는 일을 **수동으로** 흉내 내면 되겠네?
> 

`object` 대신 서비스 로케이터패턴을 사용해 `lateinit`을 제거하고, 의존성을 주입 가능하도록 변경해봅시다.

**1. `UserPreferences.kt`를 `class`로 변경**
(생성자에서 `SharedPreferences`를 주입받고, `init()`과 `lateinit`을 제거합니다.)

```kotlin
// 'object'가 아닌 'class'로 변경
class UserPreferences(private val prefs: SharedPreferences) { // (1) 생성자에서 의존성 주입
    // (2) init() 메서드 삭제
    // (3) lateinit var prefs 삭제

    fun saveUserInfo(id: String, password: String, nickname: String, mbti: String) { ... }
    fun getUserInfo() = UserInfo(...)
    fun isAutoLogin() = prefs.getBoolean(...)
    fun logout() { ... }

    // (4) companion object도 삭제 (상수들은 AppContainer나 private top-level로)
    // 이것도 논의해보면 좋은 주제 :)
}

```

**2. 의존성을 생성하고 보관하는 `AppContainer.kt` 생성**

```kotlin
class AppContainer(context: Context) {

    // (1) SharedPreferences 인스턴스 생성
    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences("spf_key", Context.MODE_PRIVATE)

    // (2) UserPreferences의 '진짜' 인스턴스 생성
    // UserPreferences는 SharedPreferences에 의존합니다.
    val userPreferences: UserPreferences = UserPreferences(prefs)

    /* // (나중에 Repository 패턴을 도입한다면)
    val authRepository: AuthRepository = AuthRepositoryImpl(userPreferences)
    val homeRepository: HomeRepository = HomeRepositoryImpl(...)
    */
}

```

**3. `DiveApplication.kt` 수정**
(Application은 `AppContainer`를 소유)

```kotlin
class DiveApplication : Application() {

    // (1) AppContainer 인스턴스를 앱 생명주기 동안 소유
    lateinit var appContainer: AppContainer
        private set // 외부에서 교체 불가능하도록 다른 조치를 취해도 좋음

    override fun onCreate() {
        super.onCreate()
        // (2) UserPreferences.init() 대신 컨테이너를 초기화
        this.appContainer = AppContainer(this)
    }
}

```

이제 `lateinit`이 제거되어 런타임 크래시 위험이 사라졌고 `UserPreferences`가 일반 `class`가 되어 ViewModel에서 테스트용 Mock을 주입할 수 있게 되었습니다 👏🏻

위의 방법 말고도 다른 방법이 있어요. 그리고 제가 생각한게 정답은 아니에요😅
그러니 지금 상황의 문제가 어떤것이고 어떻게 해결하는 방법이 있구나~로 봐주시면 좋겠습니다

nahy-512 replied:
우와,, 저는 싱글톤 관리를 위해 무조건 object를 사용해야 한다고 생각하고 있었는데 Mocking이 매우 어렵다는 단점도 있었군요..
힐트 없이 어떻게 잘 해보는 방법이 없을까 궁금했었는데 의존성을 생성하고 보관하는 클래스를 하나 더 두는 방법이 있었네요.
덕분에 많이 배워갑니다!! 추후에는 힐트 도입하더라도 지금은 알려주신 방법대로 수정해봐야겠어요. 감사합니다!!

nahy-512 replied:
1. `UserPreferences.kt`를 `class`로 변경 부분에서
> (4) companion object도 삭제 (상수들은 AppContainer나 private top-level로)

해당 코멘트를 남겨주신 이유를 여쭤봐도 될까요??
논의해보면 좋을 주제라고 말씀해주셔서, 해당 부분 관련해서 민재님의 관점이 궁금합니다!!

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,93 @@
+    private val _friendProfiles = MutableStateFlow(dummyFriendProfiles)
+    val friendProfile: StateFlow<ImmutableList<FriendProfile>> = _friendProfiles
+
+    init {
+        loadMyProfileInfo()
+    }
+
+    fun loadMyProfileInfo() {
+        _myProfile.value = UserPreferences.getUserInfo()
+    }
+
```
```kotlin
class HomeViewModel : ViewModel() {
    // ...
    init {
        loadMyProfileInfo() // (1) ViewModel 생성 시(Main Thread) 즉시 호출
    }

    fun loadMyProfileInfo() {
        // (2) UserPreferences.getUserInfo()는 SharedPreferences의
        // .getString()을 호출하는 '동기적인 디스크 I/O' 작업
        _myProfile.value = UserPreferences.getUserInfo()
    }
}

```

`SharedPreferences.getString()`은 내부적으로 XML 파일을 읽고 파싱하는 디스크 I/O 작업이에요.

1주차에서 봤었지만 DataStore를 사용하지 않은 이유가 궁금하네요🤔

어찌됐든 `init` 블록은 **메인 스레드**에서 동기적으로 디스크를 읽도록 실행돼요.

지금은 데이터가 작아 티가 안 나지만 이 작업이 조금이라도 늦어지면 UI가 멈추거나 최악은 ANR을 유발할 수 있어요. (코루틴을 사용하시니 `Dispatchers.IO`의 중요성을 잘 아실 거라 생각합니다.)

동기적으로 메인스레드에서 작업하는게 문제잖아요? 그럼 어떻게 해결할 수 있을까요?

> 동기적이 문제니까 비동기로 바꿔보고, 메인 스레드를 점유하는게 문제니까 다른 스레드로 옮겨보면 되지 않을까?
> 

I/O 작업은 `Dispatchers.IO` 스레드 풀로 보내버리면 좋겠죠! (앞서 리뷰한 `AppContainer`를 통해 `userPreferences`를 주입받았다고 가정하겠습니다.)

```kotlin
// (1) AppContainer에서 주입받도록 생성자 변경
class HomeViewModel(
    private val userPreferences: UserPreferences
) : ViewModel() {
    // ...
    init {
        loadMyProfileInfo()
    }

    fun loadMyProfileInfo() {
        viewModelScope.launch { // (2) 코루틴 시작 (기본: Main Thread)
            val userInfo = withContext(Dispatchers.IO) { // (3) I/O 스레드로 전환
                userPreferences.getUserInfo() // (4) 백그라운드에서 I/O 작업
            }
            _myProfile.value = userInfo // (5) Main 스레드로 복귀해 StateFlow 업데이트
        }
    }
}

```

(Hilt가 없으므로, Composable에서 `viewModel()`을 호출할 때 이 `userPreferences`를 주입하기 위해 `ViewModelProvider.Factory`를 커스텀해야 하는 추가 작업이 필요합니다)

어떤가요? 사실 이 문제는 DataStore를 쓰면 어느정도 간결해지기도 합니다🥲

nahy-512 replied:
우후후.. 1차 과제에서 SharedPreference를 쓰라고 되어있어서 모범생인 저로서는(?) 외면할 수 없었답니다 크크.
6차 세미나 때 코루틴을 배울 예정이라, DataStore는 아마 그때 적용해볼까 싶기도 해요!
제가 이번에 일을 여러번 하는 사람이 되어보겠습니다~~

미처 고려하지 못했던 부분까지 짚어주셔서 감사합니다 정말 많이 배워갑니다🥹

angryPodo commented on app/src/main/java/com/sopt/dive/domain/model/auth/LoginResult.kt:
```diff
@@ -0,0 +1,14 @@
+sealed class LoginResult() {
+    data object Success : LoginResult()
+    data class Failure(val errorType: LoginError) : LoginResult()
+}
```
결과를 타입화 하셨네요!👏🏻 `sealed interface`대신 `class`를 사용하신 특별한 이유가 있나요?
둘 다 하위타입을 제한하는 계층 구조를 만드는게 목적이지만 클래스기 때문에 프로퍼티나 함수를 가질 수 있어요.

때문에 성공과 실패가 공통 데이터를 가져야 한다면 class가 목적에 맞겠지만 지금은 타입을 제한하는 역할 뿐이니 interface가 더 낫다고 생각이 들어요. 🤔

나현님은 어떤 기준으로 `interface` 와 `class` 를 선택하시나요?

nahy-512 replied:
말씀주셨듯이 타입을 제한하는 역할을 수행하기 위함일 뿐이니 `interface`를 사용하는 게 맞을 것 같습니다!!
전에 과제하면서 구현했던 대로 습관적으로 `class`를 쓰면서 놓친 부분이었던 것 같습니다.
앞으로 `class`를 사용하게 된다면 저도 말씀주신 대로 프로퍼티나 함수를 가져야하는지에 대해 생각해보게 될 것 같네요:)

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/home/HomeScreen.kt:
```diff
@@ -9,36 +9,40 @@ import androidx.compose.foundation.lazy.LazyColumn
 ) {
-    val context = LocalContext.current
-    val userPreferences = remember { UserPreferences(context) }
+    val userInfo = viewModel.myProfile.collectAsStateWithLifecycle().value
```
```suggestion
    val userInfo by viewModel.myProfile.collectAsStateWithLifecycle()
```

여기서는 코틀린의 **Deligation**을 사용하는게 좋습니다ㅎㅎ
cc. @ckals413 @dmp100

nahy-512 replied:
훨씬 깔끔하고 명확해지겠네요ㅎㅎ
짚어주셔서 감사합니다! 수정하겠습니다

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/home/HomeScreen.kt:
```diff
@@ -9,36 +9,40 @@ import androidx.compose.foundation.lazy.LazyColumn
     HomeScreen(
-        myNickname = userPreferences.getUserInfo().nickname,
+        myNickname = userInfo.nickname,
+        friendList = viewModel.friendProfile.collectAsStateWithLifecycle().value,
```
상태를 수집하는 방식이 두가지네요! 수집하는 동일한 성격의 데이터지만 코드를 읽는 동료 개발자에게 해석의 리소스를 더 사용하게 만든다고 생각해요 😅

일관되게 코드를 작성하는건 어떨까요?

nahy-512 replied:
작성하면서 저도 마음에 들진 않지만 '보고 코리 달아주겠지?'하는 마음으로 내버려뒀던 건데 딱 걸렸네요~~
일관된 형태로 수정해보겠습니다!

두 경우에 대해서 민재님이 선호하시는 방법은 어떤 걸지 여쭤봐도 될까요?

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,93 @@
+import kotlinx.coroutines.flow.asStateFlow
+
+class HomeViewModel : ViewModel() {
+    private val _myProfile = MutableStateFlow(UserInfo.Fake)
+    val myProfile: StateFlow<UserInfo> = _myProfile.asStateFlow()
+
+    private val _friendProfiles = MutableStateFlow(dummyFriendProfiles)
+    val friendProfile: StateFlow<ImmutableList<FriendProfile>> = _friendProfiles
```
`myProfile`과 `friendProfiles` 두 개의 `StateFlow`를 따로 노출하는 방법을 선택하셨네요! 저는 하나의 UiState 데이터 클래스로 묶는 것이 더 좋다는 생각을 가지고 있어요.

이유는 HomeRoute가 두개의 소스를 구독하고 있어야 해요. View는 가급적 화면의 단일 상태 소스를 바라보는게 좋아요.
그리고 만약 '로딩 상태' 나 '에러 메세지'를 추가해야 한다면? `StateFlow`를 추가한다? 아니면 Profile의 Flow에 끼워넣기? 이렇게 확장성 면에서 번거로워요.

때문에 단일된 `State`를 노출시켜서 구독하는게 좋습니다ㅎㅎ ViewModel이 메서드를 실행해서 UI의 상태를 조작할때 `data class`의 `copy()` 를 사용하기도 좋겠쬬👍🏻

nahy-512 replied:
오호 하나의 UiState 데이터 클래스로 묶어야 하는 이유가 그런 부분들 때문이었군요ㅎㅎ
이런 방식으로 구현해본 덕에 필요성을 하나 더 느껴가네요!

View는 가급적 화면의 단일 상태 소스를 바라보는 게 좋은 이유에 대해서도 배워갑니다.

'로딩 상태'와 '에러 메시지'를 언급주신 이유는 아래의 경우에 대해 고려해봐야하기 때문이 맞을까요?

1. '로딩 상태' -> myProfile, friendProfile가 모두 다 정삭적으로 조회했을 때 로딩바를 해제
2. '에러 메시지' -> myProfile, friendProfile 중 하나라도 조회 실패하면 에러 다이얼로그 노출

혹시 제가 잘못 이해한 부분이 있다면 알려주시면 감사하겠습니다!

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/mypage/MyPageScreen.kt:
```diff
@@ -1,138 +1,147 @@
-
     MypageScreen(
-        userInfo = userPreferences.getUserInfo(),
+        onLogoutClick = {
+            UserPreferences.logout()
+            navigateToSignIn()
+        },
```
> Search, MyPage는 아직 뭔가 뷰모델 쓸 필요 없을 것 같아서.. SignIn 부분만 추가로 MVVM으로 바꿔볼까 합니다. 여러분들은 어떻게 생각하시나요?
> 

`MyPageRoute` 코드와 함께 질문에 답을 해볼게요. cc.@ckals413  @dmp100
저는 MVVM 디자인 패턴을 따르기로 했다면 일괄적으로 적용하는게 좋다고 생각합니다.
그렇다면 이렇게 생각해볼까요?

> "프로젝트에서 클린 아키텍쳐를 쓰기로 했지만 보일러 플레이트 코드가 과도해서 내가 만든 모듈은 구글 앱 아키텍쳐를 써야지~"
> 

어떤가요? 리팩토링 시에 신경써야 할 부분이 많아지고 일관된 코드베이스를 구축하는데 방해가 되지 않을까요?

물론 프로젝트의 목적성에 따라서 달라진다고 생각합니다만😅 일반적인 경우에서 말씀 드리고 있어요.

자 그러면 '뷰모델을 쓸 필요가 없다'라고 생각했던 `MyPage`의 코드의 구조에요.

```kotlin
@Composable
fun MyPageRoute(
    paddingValues: PaddingValues,
    navigateToSignIn: () -> Unit
) {
    // ...
    MypageScreen(
        userInfo = userPreferences.getUserInfo(), // (1) 데이터 직접 접근
        onLogoutClick = {
            UserPreferences.logout() // (2) 데이터 직접 쓰기 (Side-effect)
            navigateToSignIn() // (3) 네비게이션 로직
        },
        modifier = Modifier.padding(paddingValues)
    )
}
```

나현님이 생각하시는 Screen과 Route의 역할은 무엇인가요? 🤔

‘뷰모델이 필요 없는 간단한 화면’ 처럼 보이지만 꽤나 많은 책임이 있어요. 데이터를 읽고 사이드 이펙트를 트리거하고 네비게이션 로직을 호출하고 등등

사실 엄격하게 따지면 ViewModel이 처리해야 할 로직이에요.

MVVM 패턴에서 View는 의도를 전달하기만 해야해요. 지금은 의도가 아니라 직접 수행하고 있어요.

관심사 분리, 단일 책임 원칙에도 맞지않고 3차과제의 목적인 MVVM 적용에도 맞지 않다고 생각해요.

nahy-512 replied:
예시와 설명까지 자세히 남겨주셔서 감사합니다! 덕분에 많은 생각을 하게 되었습니다ㅎㅎ

관심사 분리와 SRP 원칙 관점에서 남겨주신 리뷰 내용에는 적극 동감합니다.

우선, 제가 남겼던
> 'Search, MyPage는 아직 뭔가 뷰모델 쓸 필요 없을 것 같아서..'

의 발언에 대한 변명부터 잠시 해보자면, 해당 리뷰를 남긴 시점에서는 `(1) 데이터 직접 접근` 코드는 있었지만 `(2)/(3)의 로그아웃 처리` 코드는 작성하기 전이라 '이정도면 안 써도 괜찮지 않을까?'라는 생각을 잠시 했었네요ㅎㅎ

추가로, 이번 3차 과제는 앱 전체에 MVVM을 적용해보는 것이 아닌, 'Home화면에 MVVM을 적용해주세요!'라는 게 과제 내용이었습니다:)
`SignIn`, `SignUp`은 제가 필요성을 느껴서 추가로 MVVM 패턴을 적용해본 거였어요.
`MyPage`까지 변경하기는 조금 귀찮았던 것도 사실입니다. 저희 합리화를 바로 지적해주시다니ㅠㅠ 역시나 날카로우십니다..

그러나 저는 아직 마이페이지는 MVVM 구조로  바꾸지 않으려고 합니다.
염려주신 두 가지 부분에 대한 제 의견을 남겨보고자 합니다!

---

> 데이터를 읽고 사이드 이펙트를 트리거하고 네비게이션 로직을 호출하고 등등</br> 사실 엄격하게 따지면 ViewModel이 처리해야 할 로직이에요.

라고 말씀해주셨는데, 저는 이 점들이 `MyPage`에서 뷰모델을 써야하는 큰 이유는 되지 않는다고 생각하긴 했습니다.

물론 뷰모델이 있다면 당연히 뷰모델이 처리할 책임이 되겠지만요..
만약 앱에서 MVC 패턴을 쓴다고 보면, 충분히 컨트롤러의 역할로 커버될 내용이라는 생각이 들었어요.

뷰모델을 쓰는 가장 큰 이유나 목적이 있다면, 저는 이를 '**ViewModel에서는 Model로부터 데이터를 요청하고 가공하여 상태로 보관하고, 이 상태를 View가 자동으로 반영하는 구조**'라고 생각했기 때문입니다.

현재 마이페이지에서 유저 정보를 변경하는 것도 아니고, '데이터 변경 → 자동 UI 반영' 의 구조가 일어날 일이 없는데 ViewModel을 새로 만들고, 코드를 옮기는 일 등등이 더 오버스펙이지 않을까?하는 생각을 했습니다.

물론 이 부분은.. 실제 프로젝트에서는 '정말 향후 수정해야할 가능성이 없을까?'를 기획 측에 물어보면서 면밀히 고려해봐야겠지만..
어쨌든 저는 솝트 과제는 어느정도 '내가 기획자다'라는 생각으로 제 의도를 최대한 담아보려고 하고 있는 중이라서요ㅎㅎ
(이랬는데 나중에 과제로 마이페이지 관련 기능들이 엄청 추가되면 눈물 흘리긴 하겠지만요..)

아무튼 필요하다고 생각되면 그때가서 고치는 방향으로 진행하지 않을까 싶어요.
지금 상황에서는 저는 `MyPage`에서 `SignIn`, `SignUp`에서만큼의 뷰모델의 필요성은 못 느끼겠어서요..! 혹시 다르게 생각하신다면 얼마든지 반박해주셔도 좋습니다☺️

> MVVM 디자인 패턴을 따르기로 했다면 일괄적으로 적용하는게 좋다고 생각합니다.

민재님이 남겨주신 이 의견과 관련해 제 생각을 추가로 남겨보자면요.
예시로 들어주신 클린 아키텍처 vs 구권아에 대해서는,, 아키텍처는 정말 큰 범위고, 혼재됐을 때 야기되는 혼란이 크다고 생각해서 저는 프로젝트에서는 일관성 있게 가져가야 한다고 생각합니다!! 이 부분은 민재님 의견에 적극 동의합니다.

다만, 아키택처 패턴의 경우 화면에서 요구하는 데이터에 로직이나 따라 유연하게 가져가보는 것도 좋다고 생각해요.

보통의 앱에는 정말 많은 화면이 있고, 어떤 화면은 정말 복잡할 수도, 어떤 화면은 정말 데이터를 표시해주는 게 끝일 정도로 단순한 경우도 있잖아요??
각 화면의 GUI와 기능명세가 다 나왔을 때, '이건 정말 이 디자인 패턴에 최적화된 화면이다' 싶다면 그걸로 가봐도 좋지 않을까 싶습니다.

MVVM을 기준으로 생각해본다면, 지나치게 기능이 없는 화면이 나올 수도, 혹은 관리해줘야 하는 상태가 정말 많은 화면이 나올 수도 있을 것 같아요. 동일한 앱에서의 화면이더라도 저는 전자라면 MVC, 후자라면 MVI 사용을 고려해볼 것 같아요.

아마 이 부분은 '이런 유연성을 허용할지'에 대해서부터가 팀원 간 합의가 선행되어야 할 사항일 것 같긴 합니다:)
그렇지만 해당 아키텍처 선정 이유에 대해 팀원들 모두가 필요성을 이해하는 상황이라면 프로젝트 전체의 일관성을 조금 버리더라도 이런 방향도 괜찮지 않을까 싶긴 해요.(겸사겸사 디자인 패턴을 정말 잘 공부할 수 있는 방법이라고도 생각합니다)
전략을 각 화면마다 다르게 가져간다면, 오히려 화면 별로 유지보수성을 오히려 높일 수 있지 않을까 하는 생각도 드네요.

물론 저도 아직까지 이런 전략을 써서 프로젝트 해본 적은 없지만ㅎㅎ 다른 분들이 해보셨다는 얘기를 듣고 재밌는 접근이라는 생각이 들어서 한번 해보고 싶었어요.

이 부분 관련해서도 민재님 의견이 정말 궁금하네요. 추가 코멘트 기다리고 있겠습니다😃
항상 여러 관점에 대해 생각해보게 해주셔서 감사합니다!!

chattymin replied:
ㅎㅎ 재밌는 이야기가 있어서 잠시 지나갑니다~

우선 저는 아키텍처와 패턴의 차이에 대해서 먼저 생각하는 편이에요.
아키텍처라는 것은 앱 전체에 영향을 끼치는 거라 앱 전반의 데이터의 흐름을 관리하는 것이기에 구글권장과 클린을 섞어쓰게 된다면 위에 말씀해주신 것처럼 일관된 코드베이스를 구축에 방해가 돼요.

하지만 패턴의 경우에는 앱 내에서 전부 통일해야 하는가? 라고 생각한다면 저는 아니라고 봅니다.
home 화면과 같이 복잡한 화면이라면 MVVM을 사용하는게 유지보수에 좋을 수 있어요.
그렇다면 setting과 같은 간단한 화면에도 MVVM을 사용해야 할까요?

조금 더 극단적으로 바라보았을 때 Network 통신이 없고, 화면 전환의 기능만 존재한다면 해당 화면도 MVVM을 사용해야 할까요?
화면 전환 event를 ViewModel 발생시키고, 이를 UI에서 collect하는게 권장되는 MVVM이니까요.

패턴의 경우에는 해당하는 컴포넌트 혹은 해당하는 레이어에만 영향을 끼쳐요.
즉, 아키텍처 패턴인 MVVM의 경우 해당하는 UI레이어에만 영향을 끼칩니다.

그렇기 때문에 각 화면의 상황에 맞는 패턴을 적용하는게 오히려 더 유연한 개발일수도 있다고 생각합니다 :)


물론 앱 전체가 하나의 패턴이 적용되어있다면 모두가 이해하기 쉽고, 하나의 일관된 목표로 달려나갈 수 있다고 생각해요
그래서 패턴을 통일하는 방법은 초기 개발시에는 선택사항으로 둘 수 있다고 생각합니다.

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/mypage/component/ProfileSpringCard.kt:
```diff
@@ -0,0 +1,179 @@
+    stiffness = stiffness,
+    mass = 1f
+)
+
+@Composable
+fun ProfileSpringCard(
+    modifier: Modifier = Modifier,
+    width: Dp = 180.dp,
+    height: Dp = 250.dp
+) {
```
요건 정말 사소한 의견이지만 앞,뒤 이미지를 받아주는 Slot을 뚫어주면 더 유연한 컴포넌트가 될 것 같네요ㅎㅎ

nahy-512 replied:
오호 이미지 슬롯을 뚫어주는 것도 좋겠네요ㅎㅎ 수정하겠습니다!

chattymin replied:
저라면 조금 더 확장해서 image가 아닌 FrontCard, BackCard 컴포저블을 받아왔을 수도 있을 것 같네요~

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/mypage/component/ProfileSpringCard.kt:
```diff
@@ -0,0 +1,179 @@
+        contentScale = ContentScale.Crop,
+        modifier = modifier
+            .offset(x = offset.dp, y = offset.dp)
+            .noRippleClickable(onClick = { onClick() })
```
이건 람다형식이 맞지 않아서 요렇게 작성했나요~?

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/mypage/component/ProfileSpringCard.kt:
```diff
@@ -0,0 +1,179 @@
+        contentDescription = null,
+        contentScale = ContentScale.Crop,
+        modifier = modifier
+            .offset(x = offset.dp, y = offset.dp)
+            .noRippleClickable(onClick = { onClick() })
+            .zIndex(if (rotation <= 90f) 1f else -1f)
+            .graphicsLayer {
+                rotationY = rotation
+            }
```
애니메이션 아주 잘 구현해주셨어요👍🏻
여기서 최적화를 약간 해보자면 Compose의 렌더링은 크게 '측정 -> 배치 -> 그리기' 단게로 이뤄져요.

`offset` 수정자는 측정과 배치 단계에서 작동해요. 즉 이걸 애니메이션 시키면 Compose는 ㅁ애니메이션의 매 프레임 마다 해당 Composable을 다시 측정하고 재배치해요. 상대적으로 비싼 작업이에요🥲

`graphicsLayer`는 그리기 단계에서 작동해요. 이 안에 있는 `translationX`, `translationY`, `rotationY`, `alpha` 등은 이미 '배치'가 끝난 Composable을 GPU를 이용해 '그리는' 단계에서만 조작해요.

그러니 애니메이션으로 변경되는 모든 값을 `graphicsLayer`에서 처리하면 좋겠죠? 그리고 '변하는 값' 을 파라미터로 전달하는것과 람다 안에서 전달하는것의 차이를 공부해보셔요ㅎㅎ

nahy-512 replied:
최적화 방향 설명해주시고, 추가로 공부할 내용까지 남겨주셔서 감사합니다! 더 알아보겠습니다ㅎㅎ

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/signin/SignInScreen.kt:
```diff
@@ -26,55 +22,44 @@ import androidx.compose.ui.text.input.ImeAction
-    var id by remember { mutableStateOf("") }
-    var password by remember { mutableStateOf("") }
+    val id = viewModel.id.collectAsStateWithLifecycle().value
+    val password = viewModel.password.collectAsStateWithLifecycle().value
```
😏

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/signin/SignInScreen.kt:
```diff
@@ -26,55 +22,44 @@ import androidx.compose.ui.text.input.ImeAction
-            } else {
-                Toast.makeText(context, "아이디 또는 비밀번호가 일치하지 않습니다.", Toast.LENGTH_SHORT).show()
+            val loginResult = viewModel.validateLogin()
+
+            when (loginResult) {
+                is LoginResult.Success -> {
+                    navigateToMain()
+                }
+                is LoginResult.Failure -> {
+                    Toast.makeText(context, loginResult.errorType.errorMessage, Toast.LENGTH_SHORT).show()
+                }
```
요게 Screen의 책임일까요~? Route에서 이펙트 핸들러를 써주면 좋아보여요잉ㅎ

nahy-512 replied:
Route에서 Screen으로 `onLoginClick` 이벤트로 넘겨주는 부분이니까 처리의 책임을 Route로 볼 수 있지 않을까 생각하긴 했습니다! SideEffect 처리에 대해서는 리팩토링 고민해보겠습니다!

angryPodo commented on app/src/main/java/com/sopt/dive/presentation/signin/SignInViewModel.kt:
```diff
@@ -0,0 +1,65 @@
+        _password.update { password }
+    }
+
+    fun validateLogin(): LoginResult {
```
`validateLogin`이라는 함수 이름을 부여 한 이상 '검증'만 해야한다고 생각해요. 지금은 검증과 사이드 이펙트 관리, 상태 저장을 하고 있어요.🤔
"함수 혹은 메서드를 한가지 일만 하도록 한다." 를 생각하면서 코드를 작성하다보면 책임에 대한 감이 올거에요ㅎㅎ
지금 같은 경우는 검증로직과 저장 로직을 분리하는게 좋겠네요

nahy-512 replied:
으흠 그러네요 처리를 한번에 해주고 싶어서 이름은 `validate`라고 지어놓고, 유저 정보 저장 로직도 넣었네요ㅠㅠ `saveUserInfo` 함수를 `private`로 만들고 싶다는 생각이 강했나 봅니다.
그럼 민재님은 함수 이름을 바꾸지 않는 한, 저장 로직은 UI 단에서 `LoginResult.Success` 경우일 때 `viewModel.saveUserInfo()`를 직접 호출하는 방법이 낫다고 보시나요?? 아니면 다른 더 좋은 방법이 있을까요?

MVI 패턴을 사용해 의도 기반으로 코드를 작성한다면 이런 고민에서 벗어날 수 있을까요..

angryPodo reviewed (commented):
오랜만에 집중해서 하느라 오래걸렸네요ㅋ
3주차 과제 고생했어요ㅎㅎ 중간중간 낚시인지 의도인지 헷갈려서 그냥 보이면 코리 달았습니다👍🏻
모르는거 있으면 언제든 물어보고, 항상 근본은 모듈화, 경계나누기를 잊지마세요! 고샌햇습니다ㅎ

nahy-512 commented:
달아주신 질문들에 대한 답변 남겨봅니다! @chattymin 

> 커밋 컨벤션을 바꾸셨는데 커밋 컨벤션의 기준이 있으신가요? (이슈 번호를 넣는다는지)

기존 커밋 내용이 가독성이 좋지 않은 것 같아서 변경해보았어요! 예전에 진행하던 프로젝트에서 쓰던 컨벤션인데, 눈에 가장 잘 들어오는 형태인 것 같아서요:)
이슈 번호는.. 어차피 번호 잘 외우지도 않기에 '굳이 제목에 들어가야하나? 오히려 보기 힘든 것 같은데?'라는 생각이 들어서 footer에서 언급하게끔 해봤습니다!

이슈 번호를 작성하는 이유는 제가 느끼기에는 이슈 내에서 어떤 커밋이 있는지 모아보기 위함.. 이 가장 컸던 것 같은데 이슈 번호를 작성하는 다른 이유도 있을까요??
다른 분들이 커밋 작성하시는 기준도 궁금하긴 하네용

nahy-512 committed:
refactor: private 접근제한자 추가

Related to: #6

nahy-512 committed:
refactor: 상태 수집에 delegate 패턴 활용

Related to: #6

nahy-512 committed:
refactor: 로그인 유효성 판단 관련 로직 수정

- LoginResult를 sealed class -> sealed interface로 변경
- 유효성 판단 책임 분리

Related to: #6

nahy-512 committed:
refactor: ProfileSpringCard 이미지, 텍스트 슬롯 추가

Related to: #6

chattymin reviewed (commented):
고생했습니다~
민재가 다 달아둬서 제가 달게 없네용 ㅋㅋ

nahy-512 committed:
refactor: onClick 람다 전달 방식 수정

Related to: #6
